### PR TITLE
PERF: BlockPlacement construction

### PR DESCRIPTION
**Pull request by jbrockmendel**
xref https://github.com/pandas-dev/pandas/pull/40171#issuecomment-790219422 when disabling fast_apply, this cuts the linked benchmark by 13%

### CODE REVIEW

File: pandas/core/indexes/range.py [hunk 1]
```diff
@@ -164,7 +164,7 @@ def _simple_new(cls, values: range, name: Hashable = None) -> RangeIndex:
         assert isinstance(values, range)
```
Comment (jbrockmendel) [inline pandas/core/indexes/range.py:164]: @simonjayhawkins are the mypy checks thorough/reliable enough that we can safely remove these assertions?

**Comment by jbrockmendel** [comment]
updated with an implementation of `BlockManager._simple_new` that brings the benchmark gains to 19-22%

**Comment by jreback** [comment]
> pandas/core/internals/managers.py:817: error: Argument 1 to "_simple_new" of "BlockManager" has incompatible type "Tuple[Any, ...]"; expected "Tuple[Block]"  [arg-type]

**Comment by jbrockmendel** [comment]
> pandas/core/internals/managers.py:817: error: Argument 1 to "_simple_new" of "BlockManager" has incompatible type "Tuple[Any, ...]"; expected "Tuple[Block]" [arg-type]

addressed + greenish

**Comment by jreback** [comment]
the numpy_dev failure could be related

**Comment by jbrockmendel** [comment]
> the numpy_dev failure could be related

im pretty sure that is from #39953 (joris commented on it and im assuming he was correct)

**Comment by jreback** [comment]
ok cool

**Pull request merged**